ext: nordic: Import nrfx 0.8.0

anangl opened the pull request:
Purpose:
   With added proper shims adapting it to Zephyr's APIs, nrfx will provide
   peripheral drivers for Nordic SoCs.

Description:
   nrfx is an extract from the nRF5 SDK that contains solely the drivers
   for peripherals present in Nordic SoCs, for convenience complemented with
   the MDK package containing required structures and bitfields definitions,
   startup files etc.

   Only relevant files from nrfx were imported, for instance, template files
   and the ones with startup code or strictly related to installation of IRQs
   in the vector table were omitted.

This pull request imports relevant files from nrfx 0.8.0 into ext/hal/nordic,
adds a proper adaptation layer for it, and finally replaces old Nordic HAL and
MDK with the refreshed ones from nrfx.

Commit:
ext: nordic: Replace HAL and MDK with the ones from nrfx

The old HAL and MDK have been removed from the source tree.
Since RADIO HAL is not yet present in nrfx, the "nrf_radio.h" file
was temporarily moved to "nrfx/hal" folder. It will be replaced with
the proper file from nrfx in its next update.

Signed-off-by: Michał Kruszewski <michal.kruszewski@nordicsemi.no>
Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
ext: nordic: Import relevant files from nrfx 0.8.0

nrfx is an extract from the nRF5 SDK that contains solely the drivers
for peripherals present in Nordic SoCs, for convenience complemented
with the MDK package containing required structures and bitfields
definitions, startup files etc.

Origin: nrfx
License: BSD 3-Clause
URL: https://github.com/NordicSemiconductor/nrfx/tree/v0.8.0
commit: b7cfe970b45ad7cc9c36b62ee620508e9e2c7fb5
Purpose: Provide peripheral drivers for Nordic SoCs
Maintained-by: External

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

carlescufi reviewed (changes requested):
>"Import relevant files from nrfx 0.8.0"

Typically we do full imports into `ext/`. Is there a reason you're not importing the full contents of the nrfx repo?

anangl commented:
@carlescufi
I thought there was no point in importing files that won't be used in Zephyr, like startup codes or IRQ vectors related. See for instance [this one](https://github.com/NordicSemiconductor/nrfx/blob/master/soc/nrfx_irqs_nrf51.h) - it renames IRQ handlers so they can be installed in the vectors table in the standard MDK way. I thought that importing it into Zephyr would only generate confusion.

carlescufi commented:
@anangl In theory `ext/` is for full, untouched imports of external modules. I will let @nashif and @galak weigh in here.

anangl commented:
Corrected commit messages.

anangl commented:
Corrected author email format in first two commits.

anangl commented:
Corrected author email address in the third commit.

SebastianBoe commented on ext/hal/nordic/Kconfig:
```diff
@@ -4,12 +4,8 @@
-
-config HAS_NORDIC_HAL
+config HAS_NORDIC_DRIVERS
 	bool
```
I'm not sure what NORDIC_DRIVERS is, but shouldn't it be selecting NRFX?

anangl replied:
Right now it is only "nrf_drv_radio802154" in ext/hal/nordic/drivers.
NRFX is selected in nrf5x/Kconfig.series. Isn't it enough, or could that be wrong?

SebastianBoe replied:
It's fine.

SebastianBoe commented on ext/hal/nordic/nrfx_config_nrf52840.h:
```diff
@@ -0,0 +1,2974 @@
+// <2=> Synth
+
+#ifndef NRFX_CLOCK_CONFIG_LF_SRC
+#define NRFX_CLOCK_CONFIG_LF_SRC 1
```
How does this configuration system integrate with Kconfig?

If it was documented to users what this file was, what would that documentation say?

anangl replied:
This file provides static configuration for nrfx drivers. This way you can tailor them to your specific needs: exclude some features or support for particular instances of peripherals, specify default values for configuration structures, etc.
Regarding integration with Kconfig - CONFIG_* macros can be used in this file if there's a need to influence the compilation of nrfx drivers. Changes to this file will come later, together with shims adapting the drivers to Zephyr's APIs.

SebastianBoe replied:
Right, so in Zephyr that use-case is covered by Kconfig. Users should not need to learn a new mechanism to be able to configure a new subsystem in Zephyr. This does not have to block the merge, but eventually we will need to have the ability to configure every single on of these configs through Kconfig.

I suspect you will be able to extend your nrf_config_*.h generator to also be able to generate a Kconfig file. After you have that in place it might be possible to have a simple adapter that does

#define NRFX_CLOCK_CONFIG_LF_SRC CONFIG_NRFX_CLOCK_CONFIG_LF_SRC

for each config.

SebastianBoe reviewed (approved):
LGTM wrt. the build system

MaureenHelm commented on ext/hal/nordic/CMakeLists.txt:
```diff
@@ -11,3 +11,9 @@ if(CONFIG_HAS_NORDIC_DRIVERS)
   add_subdirectory(drivers)
 endif()
+
+if(CONFIG_HAS_NRFX)
```
Is there a reason this change and the kconfig changes aren't in the glue logic commit instead?

anangl replied:
Actually, there isn't.
I'll move these changes to the next commit, as it seems to be a better place for them, indeed.

MaureenHelm commented on ext/hal/nordic/nrfx_config.h:
```diff
@@ -0,0 +1,47 @@
+ * All rights reserved.
+ * 
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
```
Should this file have been added in the import commit?

anangl replied:
I don't think do. This file was not imported, it was created specially for Zephyr. nrfx provides only templates of configuration files.

MaureenHelm commented on ext/hal/nordic/nrfx_config_nrf51.h:
```diff
@@ -0,0 +1,1583 @@
+#ifndef NRFX_CONFIG_NRF51_H__
```
Need to add a license header

anangl replied:
Right. I was considering the "nrfx_config_*.h" files as parts of the "nrfx_config.h" file, which has a proper license header. But I think it was not a good approach. I'll add the headers.

MaureenHelm commented on arch/arm/soc/nordic_nrf5/nrf51/Kconfig.series:
```diff
@@ -16,8 +16,6 @@ config  SOC_SERIES_NRF51X
 	select SYS_POWER_LOW_POWER_STATE_SUPPORTED
 	select XIP
 	select HAS_CMSIS
-	select HAS_NORDIC_MDK
```
The commit message should mention removing the HAL/MDK

anangl replied:
You mean "Replace HAL and MDK..." in the subject line is not enough? Should I add something like "The old HAL and MDK have been removed from the source tree." in the message body?

MaureenHelm replied:
When I do something like this, I usually like to do it in three steps: 1) add the new thing 2) rework  plumbing to use the new thing 3) remove the old thing

So, when I first read the commit message, I thought you were just doing 2)

MaureenHelm reviewed (commented):
> @anangl In theory ext/ is for full, untouched imports of external modules. I will let @nashif and @galak weigh in here.

ext/ is for untouched imports, but not necessarily full. We've only been importing the parts that we need.

anangl committed:
ext: nordic: Add initial adaptation layer for nrfx

This commit adds a glue layer that adapts nrfx to be usable in Zephyr
as a host environment and files with static configuration of nrfx
drivers for several supported SoCs.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
Addressed review comments.

carlescufi reviewed (changes requested):
>nrfx is an extract from the nRF5 SDK that contains solely the drivers
for peripherals present in Nordic SoCs, for convenience complemented
with the MDK package containing required structures and bitfields
definitions, startup files etc.

I don't think I would call `nrfx` an "extract". I would instead say something like "nrfx is a standalone driver layer for Nordic nRF SoCs"

carlescufi reviewed (changes requested):
>This commit adds gluing layer that

"This commit adds a glue layer that..."

anangl commented:
@carlescufi Regarding the "extract", I took this block of text from the [nrfx documentation](https://github.com/NordicSemiconductor/nrfx/blob/master/doc/main_page.dox#L4). But of course I can rephrase it if you find it inappropriate. The same block I used in [README](https://github.com/anangl/zephyr/blob/752f0d063d932a2f7cf2c6b791d199b02446b5b6/ext/hal/nordic/nrfx/README#L15).

carlescufi commented:
@anangl OK, let's modify it upstream in nrfx and then use the new text in the next import.

Can you change the commit message I pointed? s/gluing/glue

anangl commented:
@carlescufi Done.

carlescufi commented:
Unless someone has further comments this should be ready to merge
